### ISSUE TITLE
Accessing HttpRequestData in middleware

### ISSUE THREAD

**Issue by hajekj** [Azure/azure-functions-dotnet-worker#414]
As per the [sample](https://github.com/Azure/azure-functions-dotnet-worker/blob/main/samples/CustomMiddleware/MyCustomMiddleware.cs) I am able to build a custom middleware which accesses the `FunctionContext` and works with it. However, how do I access the `HttpRequestData` object?

Some broader context: I am trying to build an authentication middleware: https://hajekj.net/2022/04/22/azure-functions-out-of-process-and-authentication-with-azure-ad/

I wanted to do it via `context.Features.GetRequired<IFunctionBindingsFeature>();` but [`IFunctionBindingsFeature` is internal](https://github.com/Azure/azure-functions-dotnet-worker/blob/72b9d17a485eda1e6e3626a9472948be1152ab7d/src/DotNetWorker.Core/Context/Features/IFunctionBindingsFeature.cs#L9). Another approach would be to go via the raw data: `context.BindingContext.BindingData;` but that would require me to manually parse the headers JSON.

So far, I managed to parse out the headers and seems to work fine - but it looks really dirty to me - https://github.com/hajekj/azure-functions-dotnet-worker-miw/blob/master/WorkerAuthentication/AuthenticationMiddleware.cs.

Also, I would like to be able to modify the response - for example to return HTTP 401 when token is not present, but I didn't manage to figure out how to do it in middleware. Is there any way to do it - or modify the Function output in general?

**Comment by godefroi** [comment]
> Also, I would like to be able to modify the response - for example to return HTTP 401 when token is not present, but I didn't manage to figure out how to do it in middleware. Is there any way to do it - or modify the Function output in general?

This isn't currently possible. See #340

**Comment by david-peden-q2** [comment]
While waiting for a proper solution, this works:

```csharp
public static class FunctionContextExtensions
{
    public static HttpRequestData GetHttpRequestData(this FunctionContext functionContext)
    {
        try
        {
            KeyValuePair<Type, object> keyValuePair = functionContext.Features.SingleOrDefault(f => f.Key.Name == "IFunctionBindingsFeature");
            object functionBindingsFeature = keyValuePair.Value;
            Type type = functionBindingsFeature.GetType();
            var inputData = type.GetProperties().Single(p => p.Name == "InputData").GetValue(functionBindingsFeature) as IReadOnlyDictionary<string, object>;
            return inputData?.Values.SingleOrDefault(o => o is HttpRequestData) as HttpRequestData;
        }
        catch
        {
            return null;
        }
    }
}
```

And you can use it like this:

```csharp
public class CustomMiddleware : IFunctionsWorkerMiddleware
{
    public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
    {
        HttpRequestData httpRequestData = context.GetHttpRequestData();

        // do something with httpRequestData

        await next(context);
    }
}
```

**Comment by fabiocav** [comment]
Thank you for opening the issue and the workaround shared.

Currently, as you've found, the context carries the raw information and the services that would enable binding processing are not publicly exposed. This is work that is planned and, although I don't have a concrete ETA, we hope to expose soon. Keeping this open so we can update with progress.

**Comment by nlpMbordogna** [comment]
Any change of a work around for the missing OutputData, something we can reflect to retrieve the HttpResponseData class?  I inspected the data at runtime and cannot find anything similar

**Comment by rhollamby** [comment]
Yes I am also struggling with this. I need to set some http headers for all requests. Even if I create a service that I inject into each function it's still really messy.

**Comment by JanosNollFD** [comment]
> Any change of a work around for the missing OutputData, something we can reflect to retrieve the HttpResponseData class? I inspected the data at runtime and cannot find anything similar

The workaround is so hacky it almost shouldn't be used, but here goes:
```
var feature = context.Features.FirstOrDefault(f => f.Key.Name == "IFunctionBindingsFeature").Value;
if (feature == null) throw new Exception("Cannot get function bindings feature");
PropertyInfo pinfo = feature.GetType().GetProperty("InvocationResult");
HttpResponseData resp = ...
// create a new instance of a child class because HttpResponseData is abstract. You can set 
// StatusCode, Body, HeadersCollection, but for some reason Body doesn't work (content not sent to the calling client)
pinfo.SetValue(feature, resp);
```

I hope the proper interfaces get exposed publicly soon.

**Comment by chandsalam1108** [comment]
Hi, How can we get access to HttpContext object to set User Claims?

In my case I'm using custom Authorization lib to validate JWT token. Upon successful validation I need to set user claims to HttpContext so that I can use that in other parts of function app method execution.

Thank you!
Chand

**Comment by arashcomsense** [comment]
@ThomasArdal I tried [`GetHttpResponseData()`](https://github.com/elmahio/Elmah.Io.Functions.Isolated/blob/065090253ff62d9eda54d6a00204996bd5f94235/src/Elmah.Io.Functions.Isolated/FunctionContextExtensions.cs#L33-L51) but the issue is that it overwrites the already-generated response object in the middleware executed after the function. For e.g. in my case, I need to add a header to every single  function response. Is there a way to be able to manipulate the "existing" `HttpResponseData` object?

**Comment by ThomasArdal** [comment]
@arashcomsense Don't know. I think you should use StackOverflow or create another issue since that is out of the scope of this issue.

**Comment by SeanFeldman** [comment]
As someone that has gone through trying to figure out how to do authentication using middleware for this SDK, this is not the experience I'd like developers to go through. Isolated Worker SDK middleware is a solid tool to tap into the pipeline. It falls short by supporting anything other than the trivial termination or enrichment. Combined with the forced reflection to access specifically typed needed at runtime, it cast a massive shadow on the maturity of this SDK despite the GA label.

More samples would help but would likely add the burden on the team to maintain those. Not to mention the hesitation to show how to do things in a ["wrong" way](https://github.com/Azure/azure-functions-dotnet-worker/issues/414#issuecomment-872818004). Support for [SDK types](https://github.com/Azure/azure-functions-dotnet-worker/issues/226), [imperative bindings](https://github.com/Azure/azure-functions-dotnet-worker/issues/303), [Application Insights](https://github.com/Azure/azure-functions-dotnet-worker/issues/760), access to Functions SDK types, all these issues point to the same root problem - missing clarity on the path going forward for this SDK. Not to mention the [convergence of the two SDKs](https://github.com/Azure/azure-functions-dotnet-worker/issues/226#issuecomment-1092224827) into one after .NET 7.

**Comment by Arash-Sabet** [comment]
@SeanFeldman Very well said.

**Comment by kshyju** [comment]
You can use the `GetHttpRequestDataAsync` extension method to get the `HttpRequestData` instance. This API is available from [Version 1.8.0-preview1](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker/) onwards.

````
HttpRequestData requestData = await functionContext.GetHttpRequestDataAsync();
````

Please refer the ` StampHttpHeaderMiddleware` middleware in the[ sample app](https://github.com/Azure/azure-functions-dotnet-worker/blob/main/samples/CustomMiddleware/StampHttpHeaderMiddleware.cs#L15) for usage of this API in a middleware.

Thanks for your patience. Please let us know if you have questions.
